pkg/mutation/mutators/assignimage/imageparser.go

```diff
@@ -0,0 +1,114 @@
+package assignimage
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+type image struct {
+	domain string
+	path   string
+	tag    string
+}
+
+var (
+	// domainRegexp defines a schema for a domain component. Primarily, this is to
+	// avoid the possibility of injecting `/` tokens, which could cause part of
+	// the domain component to "leak" over to the location component, ultimately
+	// causing non convergence.
+	domainRegexp = regexp.MustCompile(`^\w[\w.\-_]*(:\d+)?$`)
+
+	// pathRegexp defines a schema for a location component. It follows the convention
+	// specified in the docker distribution reference. The regex  restricts
+	// location-components to start with an alphanumeric character, with following
+	// parts able to be separated by a separator (one period, one or two
+	// underscore and multiple dashes).
```

Comment:
Where does the one or two underscore rule come from? Is more than two underscores disallowed?

Reply:
This comes from the regex docker uses to validate the path: https://github.com/docker/cli/blob/master/vendor/github.com/docker/distribution/reference/regexp.go#L15-L21

The docker code uses this small library that lets you compose regex out of smaller regexes, which makes the code readable but the resulting regex is generally not comprehensible to humans (very long, many nested groups, ect). Since we show this regex as part of the error message when validation fails, I thought it best to use our own regex that tries to be simpler.

Regarding the underscores: the regex we use is a subset of what docker is doing for the aforementioned reason. We make sure no "splitting tokens" are injected and do some other easy validation but not everything docker does. A triple underscore like "my___repo/app" would technically be allowed here, but IMO this is better than overcomplicating the regex.

Reply:
Yeah, we aren't necessarily concerned with full docker parsing/validation, we are concerned with correctly identifying the three components of an image name (host, path, tag/digest)